Anomaly expansion - part 2 - Delimber anomaly

Ghilker opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
A new anomaly is here! the Delimber anomaly!
contrary to its name, this anomaly doesn't do you dirty by removing all your limbs.
It will instead improve your body by randomizing your limbs with every species known! Isn't that amazing?!
Incredibly enough it can also change your organs, improving them and expanding them! (we think this is amazing!)
Now you can also harness that power by constructing a reactive armor with its core, but be aware to not get hit by an EM pulse, our insurance doesn't cover that. 
A note on safety, this anomaly doesn't stop, will keep on pulsating until eliminated.
<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
More non grief anomalies to add to the pool, expanding the amount of interesting anomalies that can spawn around the station
<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog

<!-- If your PR modifies aspects of the game that can be concretely observed by players or admins you should add a changelog. If your change does NOT meet this description, remove this section. Be sure to properly mark your PRs to prevent unnecessary GBP loss. You can read up on GBP and it's effects on PRs in the tgstation guides for contributors. Please note that maintainers freely reserve the right to remove and add tags should they deem it appropriate. You can attempt to finagle the system all you want, but it's best to shoot for clear communication right off the bat. -->

:cl:
expansion: A new anomaly is here! the Delimber anomaly! contrary to its name, this anomaly doesn't do you dirty by removing all your limbs. It will instead improve your body by randomizing your limbs with every species known! Isn't that amazing?! Don't get close to it while wearing a biosuit or you'll lose all of its beneficial aspects! (biosuit prevents your limbs from being randomized, also anything with 100 bio armor will do). Incredibly enough it can also change your organs, improving them and expanding them! (ANY organs available are in the pool) Now you can also harness that power by constructing a reactive armor with its core, but be aware to not get hit by an EM pulse, our insurance doesn't cover that. 
expansion: A note on safety, this anomaly doesn't stop, will keep on pulsating until eliminated.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Ghilker committed:
delimber anomaly

TheBonded commented:
perfect

Pickle-Coding commented:
Can it give you plasmaman limbs?

Ghilker committed:
use replace limb

Ghilker commented:
> Can it give you plasmaman limbs?

yes, but you don't burn

Colovorat commented:
does plasmeme suit protects you from it?

GoldenAlpharex commented:
For the changelog entry, I think what you meant to say instead of 
> Don't get close to it **by using** a biosuit **of** you'll lose all of its beneficial aspects! (biosuit prevents your limbs from being randomized)

is this:
> Don't get close to it **while wearing** a biosuit **or** you'll lose all of its beneficial aspects! (biosuit prevents your limbs from being randomized)

MrMelbert commented on code/modules/clothing/suits/bio.dm:
```diff
@@ -4,7 +4,7 @@
 	desc = "A hood that protects the head and face from biological contaminants."
 	permeability_coefficient = 0.01
-	clothing_flags = THICKMATERIAL | BLOCK_GAS_SMOKE_EFFECT | SNUG_FIT | PLASMAMAN_HELMET_EXEMPT
+	clothing_flags = THICKMATERIAL | BLOCK_GAS_SMOKE_EFFECT | SNUG_FIT | PLASMAMAN_HELMET_EXEMPT | BIOHAZARDS_SAFE
 	armor = list(MELEE = 0, BULLET = 0, LASER = 0,ENERGY = 0, BOMB = 0, BIO = 100, FIRE = 30, ACID = 100)
```
Why add a whole new flag instead of using `BIO` armor?

MrMelbert commented on code/modules/mob/living/carbon/carbon_defines.dm:
```diff
@@ -33,6 +33,7 @@
 	var/obj/item/clothing/neck/wear_neck = null
 	var/obj/item/tank/internal = null
 	var/obj/item/clothing/head = null
+	var/obj/item/clothing/wear_suit = null
```
Seems very sus to move this to `/carbon` entirely instead of just casting to `/human` when you're checking for it, or instead of using `get_item_by_slot()` to get it

Ghilker replied:
It's to allow the anomaly to also work on monkeys, is there any other way to do it?

MrMelbert replied:
Like I said, either:
- cast to human
  - `if(ishuman(affected mob)); var/mob/living/carbon/human/affected_human = affected mob; affected_human.wear_suit`, etc
- use `get_item_by_slot(ITEM_SLOT_OCLOTHING)`, as it's a mob level proc

Acer202 commented:
Why not "re-limber"? More intuitive and has pun potential, ohoh

Shadowflame909 commented:
EVERY species? or just the roundstart ones.

Getting shadow-person limbs would be quite the round-changing nerf.

While golem limbs would aesthetically look cool.

Ghilker committed:
immortal anomaly,
fix some issue,
requested changes,
anomaly now changes your organs too,
reactive armor does that too

Ghilker commented:
> EVERY species? or just the roundstart ones.
> 
> Getting shadow-person limbs would be quite the round-changing nerf.
> 
> While golem limbs would aesthetically look cool.

yes every species, anything we added that has limbs. 
Also the anomaly can now change your organs too

Ghilker committed:
announce

TheBonded commented:
maybe make the anomalys range slowly increase over time, or allow it to phase through walls, so you have incentive to actually defeat it instead of turning it into a maints tourist attraction

MrMelbert commented on code/game/objects/effects/anomalies.dm:
```diff
@@ -419,4 +421,79 @@
+	for(var/mob/living/carbon/nearby in range(swap_range, src))
+		var/obj/item/clothing/suit/wear_suit = nearby.get_item_by_slot(ITEM_SLOT_OCLOTHING)
+		var/obj/item/clothing/head/wear_helmet = nearby.get_item_by_slot(ITEM_SLOT_HEAD)
+		if(wear_suit?.armor[BIO] == 100 && wear_helmet?.armor[BIO] == 100)
```
If you wanna check them directly for their bio armor values, you should use `getRating`.
(Also just suggesting that they're changed to `>=` instead of `==`, just for sanity reasons.)
```suggestion
		if(wear_suit?.armor.getRating(BIO) >= 100 && wear_helmet?.armor.getRating(BIO) >= 100)
```

BUT there's a slightly easier option which you can just run `getarmor(type = BIO)` on the mob itself to get how much total armor of a certain type the mob has 
```suggestion
		if(nearby.getarmor(type = BIO) >= 100)
```

You can even go one layer deeper and use `run_armor_check()` to include a message displayed if the mob's protected. 
```suggestion
		if(nearby.run_armor_check(type = BIO, absorb_text = "Your armor protects you from [src]!") >= 100)
```

Ghilker replied:
Ohh I like that last one, does it work on base carbons too or only on carbon/human types?

MrMelbert replied:
run_armor_check is a `living` level proc, so it'd work here

Ghilker committed:
requested change

private-tristan commented:
at fist I thought this was just kinda meh, but if the reactive armor still does what the anomoly does, then I can forsee people having lots of !!FUN!! with it

also is it able to give you implants/new organs? (I.E  etheralrevival crystal as a human, cybernetic eyes)

Ghilker commented:
will fix the conflicts after #66392 is merged

Ghilker committed:
merge master - fix

Ghilker committed:
icon fix

Kapu1178 commented on code/game/objects/effects/anomalies.dm:
```diff
@@ -419,6 +421,82 @@
+	var/static/list/r_legs
+	var/static/list/organs
+
+/obj/effect/anomaly/delimber/Initialize(mapload, new_lifespan, drops_core)
+	. = ..()
+	if(!chests)
+		chests = subtypesof(/obj/item/bodypart/chest)
+	if(!heads)
+		heads = subtypesof(/obj/item/bodypart/head)
+	if(!l_arms)
+		l_arms = subtypesof(/obj/item/bodypart/l_arm)
+	if(!r_arms)
+		r_arms = subtypesof(/obj/item/bodypart/r_arm)
+	if(!l_legs)
+		l_legs = subtypesof(/obj/item/bodypart/l_leg)
+	if(!r_legs)
+		r_legs = subtypesof(/obj/item/bodypart/r_leg)
+	if(!organs)
+		organs = subtypesof(/obj/item/organ)
```
You can use `typesof()`, all of the basetypes are human limbs

Kapu1178 commented on code/modules/clothing/suits/reactive_armour.dm:
```diff
@@ -373,3 +374,85 @@
+	var/static/list/r_legs
+	var/static/list/organs
+
+/obj/item/clothing/suit/armor/reactive/delimbering/Initialize(mapload)
+	. = ..()
+	if(!chests)
+		chests = subtypesof(/obj/item/bodypart/chest)
+	if(!heads)
+		heads = subtypesof(/obj/item/bodypart/head)
+	if(!l_arms)
+		l_arms = subtypesof(/obj/item/bodypart/l_arm)
+	if(!r_arms)
+		r_arms = subtypesof(/obj/item/bodypart/r_arm)
+	if(!l_legs)
+		l_legs = subtypesof(/obj/item/bodypart/l_leg)
+	if(!r_legs)
+		r_legs = subtypesof(/obj/item/bodypart/r_leg)
```
does all this really need to be copypasted?

Anyways, `typesof()`

Kapu1178 commented on code/game/objects/effects/anomalies.dm:
```diff
@@ -419,6 +421,82 @@
+		if(nearby.run_armor_check(attack_flag = BIO, absorb_text = "Your armor protects you from [src]!") >= 100)
+			continue //We are protected
+		var/picked_zone = pick(zones)
+		var/obj/item/bodypart/picked_user_part = nearby.get_bodypart(picked_zone)
```
get_bodypart() can return null if the target is missing said bodypart.

Ghilker replied:
i'll add a check later

Ghilker committed:
reviews

Ghilker committed:
merge master

tralezab commented on code/game/objects/effects/anomalies.dm:
```diff
@@ -421,6 +423,83 @@
+	icon_state = "delimber_anomaly"
+	aSignal = /obj/item/assembly/signaler/anomaly/delimber
+	immortal = TRUE
+	/// Time passed since the last effect, increased by delta_time of the SSobj
+	var/ticks = 0
+	/// How many seconds between each anomaly pulses
+	var/releasedelay = 15
```
I think a cooldown would be better but this is fine, as long as you convert it to deciseconds

tralezab commented on code/game/objects/effects/anomalies.dm:
```diff
@@ -421,6 +423,83 @@
+	aSignal = /obj/item/assembly/signaler/anomaly/delimber
+	immortal = TRUE
+	/// Time passed since the last effect, increased by delta_time of the SSobj
+	var/ticks = 0
```
see my problem is that release delay is in seconds, ticks is in deciseconds, and ticks doesn't properly describe itself but is rather only describing the measurement of time

tralezab commented on code/game/objects/effects/anomalies.dm:
```diff
@@ -421,6 +423,83 @@
+		var/obj/item/organ/new_organ = new picked_organ
+		new_organ.Insert(nearby, TRUE, FALSE)
+		nearby.update_body(TRUE)
+		to_chat(nearby, span_notice("Something has changed about you!"))
```
this would be so much better as a balloon alert!

tralezab commented on code/modules/clothing/suits/reactive_armour.dm:
```diff
@@ -373,3 +374,85 @@
+		var/obj/item/organ/new_organ = new picked_organ
+		new_organ.Insert(nearby, TRUE, FALSE)
+		nearby.update_body(TRUE)
+		to_chat(nearby, span_notice("Something has changed about you!"))
```
this would be so much better as a balloon alert!

tralezab reviewed (changes requested):
i'm here. i'm here to complain about time

Ghilker committed:
review

tralezab commented on code/game/objects/effects/anomalies.dm:
```diff
@@ -421,6 +423,84 @@
+		var/obj/item/organ/new_organ = new picked_organ
+		new_organ.Insert(nearby, TRUE, FALSE)
+		nearby.update_body(TRUE)
+		balloon_alert(nearby, "Something has changed about you!")
```
```suggestion
		balloon_alert(nearby, "something has changed about you")
```

tralezab commented on code/modules/clothing/suits/reactive_armour.dm:
```diff
@@ -373,3 +374,86 @@
+		var/obj/item/organ/new_organ = new picked_organ
+		new_organ.Insert(nearby, TRUE, FALSE)
+		nearby.update_body(TRUE)
+		balloon_alert(nearby, "Something has changed about you!")
```
```suggestion
		balloon_alert(nearby, "something has changed about you")
```

tralezab reviewed (approved):
you forgot to change the balloon alerts to be worded like em (no caps, ! is a fail condition). I'll just commit for you dw

tralezab committed:
Apply suggestions from code review
